dataforest/core/PlotWidget.py

```diff
@@ -0,0 +1,83 @@
+from copy import deepcopy
+from typing import TYPE_CHECKING, Dict, Any
+
+from IPython.display import Image
+import ipywidgets as widgets
+from matplotlib.figure import Figure
+
+from dataforest.structures.cache.PlotCache import PlotCache
+
+if TYPE_CHECKING:
+    from dataforest.core.DataBranch import DataBranch
+
+
+class PlotWidget:
+    def __init__(self, tree, plot_key, use_saved=True, **plot_kwargs):
+        self._tree = tree
+        self._branch_spec = deepcopy(tree.tree_spec.branch_specs[0])
+        self._plot_key = plot_key
+        self._use_saved = use_saved
+        self._plot_kwargs = plot_kwargs
+        self._sweeps = self._tree.tree_spec.sweep_dict
+        self._plot_cache = dict()
+        # self.ax = plt.gca()
+
+    def control(self):
+        process = self._tree.current_process
+        precursors = self._tree.tree_spec.get_precursors_lookup(incl_current=True)[process]
+        sweeps = set.union(*[self._sweeps[precursor] for precursor in precursors])
+        param_sweeps = {":".join(swp[:3][::-1]): list(swp[3]) for swp in sweeps}
+        _kwargs = {**param_sweeps, **self._plot_kwargs}
+
+        @widgets.interact(**_kwargs)
+        def _control(**kwargs: Dict[str, Any]):
+            for param_keys_str, value in kwargs.items():
+                if param_keys_str in self._plot_kwargs:
+                    self._plot_kwargs[param_keys_str] = value
+                    continue
+                elif isinstance(value, float):
+                    value = int(value) if int(value) == value else value
+                (name, operation, process) = param_keys_str.split(":")
+                self._branch_spec[process][operation][name] = value
+            # TODO: do we need to do something with kwargs? Or taken care of?
+            [kwargs.pop(k) for k in self._plot_kwargs]
+            branch = self._tree._branch_cache[str(self)]
+            return self._get_plot(branch)
+
+        return _control
+
+    def _get_plot(self, branch):
+        plot_map = branch[self._tree.current_process].plot_map
+        plot_path_lookup = {plot_key: next(iter(path_dict.values())) for plot_key, path_dict in plot_map.items()}
+        spec = branch.spec[: branch.current_process]
+        cache_key = str({"spec": spec, "plot_kwargs": self._plot_kwargs})
+        if self._plot_key in plot_path_lookup and self._use_saved:
+            plot_path = plot_path_lookup.get(self._plot_key)
+            if plot_path.exists():
+                return Image(plot_path)
+        generated = False
+        if not (plot_obj := self._plot_cache.get(cache_key, None)):
```

Comment:
Haha, I didn't know about the `:=` operator :)

Reply:
Oh shoot, that's the walrus from python 3.8 (more dope operators coming in 3.9!), but I had intended to support 3.6+, so I should probably remove that. I think I've done some other naughty stuff elsewhere, so at some point, I'll do a compatibility sweep on 3.6 and fix everything